packages/plugins/other/visitor-plugin-common/src/base-resolvers-visitor.ts

```diff
@@ -1306,6 +1302,33 @@ export class BaseResolversVisitor<
       ).string;
   }
 
+  public buildFederationReferenceTypes(): string {
+    const federationMeta = this._federation.getMeta();
+
+    if (Object.keys(federationMeta).length === 0) {
+      return '';
+    }
+
+    const declarationKind = 'type';
+    return new DeclarationBlock(this._declarationBlockConfig)
+      .export()
+      .asKind(declarationKind)
+      .withName(this.convertName('FederationReferenceTypes'))
+      .withComment('Mapping of federation reference types')
+      .withBlock(
+        Object.entries(federationMeta)
+          .map(([typeName, { referenceSelectionSetsString }]) => {
+            if (!referenceSelectionSetsString) {
+              return undefined;
+            }
+
+            return indent(`${typeName}: ${referenceSelectionSetsString}${this.getPunctuation(declarationKind)}`);
+          })
+          .filter(v => v)
+          .join('\n')
+      ).string;
+  }
```

Comment:
Having `buildFederationReferenceTypes` to build standalone `FederationReferenceTypes` allows us to easily refer to the complex `reference` param types in different scenarios:
- When using mappers:
```ts
import type { FederationReferenceTypes } from "./types.generated";

export type UserMapper = DatabaseUser | FederationReferenceTypes["User"]
```